fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1301,15 +1301,19 @@ def _remove_shard_bwd_hook(fsdp_module: FullyShardedDataParallel) -> None:
             if isinstance(m, FullyShardedDataParallel):
                 _remove_shard_bwd_hook(m)
                 m._pre_backward_hook_has_run = False
-                if m._has_params:
-                    if any(p.requires_grad for p in m.params):
+                # Note: m.parameters() should not be an empty list. FSDP
+                # wrapping modules without weights is not tested at the moment.
+                if any(p.requires_grad for p in m.parameters()):
+                    if m._has_params:
```

Comment:
what does `_has_params`  bool signify?

Reply:
It indicates if this instance has any “owned” params or not. It is possible that all params within this module is owned by children FSDP instances, leaving the parent empty without any owned params